### PR TITLE
Null qubit preprocessing

### LINKED ISSUE #5508: [BUG] `null.qubit` no longer plays nicely with custom operations
### Expected behavior

``` py
import pennylane as qml
from pennylane.resource.error import ErrorOperation

class IssueErrorOp(ErrorOperation):
    def __init__(self, phase, wires):
        self.phase = phase
        super().__init__(phase, wires=wires)
    def error(self):
        return 1.0

@qml.qnode(qml.device('null.qubit', wires=2))
def circuit():
    IssueErrorOp(1.0, [0])
    return qml.state()

>>> circuit()
[0.0]
```

`null.qubit` should allow for adding custom operators to the workflow.

### Actual behavior

``` py
import pennylane as qml
from pennylane.resource.error import ErrorOperation

class IssueErrorOp(ErrorOperation):
    def __init__(self, phase, wires):
        self.phase = phase
        super().__init__(phase, wires=wires)
    def error(self):
        return 1.0

@qml.qnode(qml.device('null.qubit', wires=2))
def circuit():
    IssueErrorOp(1.0, [0])
    return qml.state()

>>> circuit()
DeviceError: Operator ReplyErrorOp(1.0, wires=[0]) not supported on null.qubit and does not provide a decomposition.
```

### Additional information

This issue exists on the master and possibly comes from using `DefaultQubit`'s `preprocess` for `null.qubit` as well. 

### Source code

_No response_

### Tracebacks

_No response_

### System information

```shell
Name: PennyLane
Version: 0.36.0.dev0
Summary: PennyLane is a cross-platform Python library for quantum computing, quantum machine learning, and quantum chemistry. Train a quantum computer the same way as a neural network.
Home-page: https://github.com/PennyLaneAI/pennylane
Author: 
Author-email: 
License: Apache License 2.0
Location: /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages
Editable project location: /Users/utkarsh/Xanadu/Codes/SourceCodes/pennylane
Requires: appdirs, autograd, autoray, cachetools, networkx, numpy, pennylane-lightning, requests, rustworkx, scipy, semantic-version, toml, typing-extensions
Required-by: amazon-braket-pennylane-plugin, PennyLane-Catalyst, PennyLane-Cirq, PennyLane-qiskit, PennyLane_Lightning, quantum-datasets

Platform info:           macOS-13.6.4-arm64-arm-64bit
Python version:          3.10.4
Numpy version:           1.26.4
Scipy version:           1.12.0
Installed devices:
- default.clifford (PennyLane-0.36.0.dev0)
- default.gaussian (PennyLane-0.36.0.dev0)
- default.mixed (PennyLane-0.36.0.dev0)
- default.qubit (PennyLane-0.36.0.dev0)
- default.qubit.autograd (PennyLane-0.36.0.dev0)
- default.qubit.jax (PennyLane-0.36.0.dev0)
- default.qubit.legacy (PennyLane-0.36.0.dev0)
- default.qubit.tf (PennyLane-0.36.0.dev0)
- default.qubit.torch (PennyLane-0.36.0.dev0)
- default.qutrit (PennyLane-0.36.0.dev0)
```

### Existing GitHub issues

- [X] I have searched existing GitHub issues to make sure the issue does not already exist.

### PR DESCRIPTION
**Pull request by albi3ro**
**Context:**

The new and improved `null.qubit` strictly inherits it's preprocessing from `default.qubit`. We need a bit more flexibility than that though for tracking error propagation and resources and examining the classical overhead associated with arbitrary plugin devices.

**Description of the Change:**

Adds three new keyword arguments:

1) `target_device`.  This accepts a device implementing the new device interface. `null.qubit` will then get it's preprocessing from that device.

3) `assume_no_broadcasting`.  When I was using `null.qubit` for profiling, I noticed a huge chunk of time was spent calculating the batch size to determine the proper shape for the measurement.  This skews the profiling numbers in a way we don't want.  The easiest option is simply turning off that calculation.

**Benefits:**

`null.qubit` is more versatile and useful.

**Possible Drawbacks:**

Too many keyword arguments.

**Related GitHub Issues:**

Fixes #5508 [sc-61258] [sc-57971]

### CODE REVIEW

File: doc/releases/changelog-dev.md [hunk 1]
```diff
@@ -266,6 +266,9 @@
 * Removed the warning that an observable might not be hermitian in `qnode` executions. This enables jit-compilation.
   [(#5506)](https://github.com/PennyLaneAI/pennylane/pull/5506)
 
+* `null.qubit` can now accept a target gateset via `operations`, a device to mimic via `target_device`,
+  and will assume no parameter broadcasting if `assume_no_broadcasting=True`.
```
Comment (albi3ro) [inline doc/releases/changelog-dev.md:270]:
```suggestion
  and will assume no parameter broadcasting if `assume_no_broadcasting=True`.
    [(#5527)](https://github.com/PennyLaneAI/pennylane/pull/5527)
```

File: pennylane/devices/null_qubit.py [hunk 2]
```diff
@@ -137,6 +139,10 @@ class NullQubit(Device):
             (``['aux_wire', 'q1', 'q2']``). Default ``None`` if not specified.
         shots (int, Sequence[int], Sequence[Union[int, Sequence[int]]]): The default number of shots
             to use in executions involving this device.
+        operations (Optional[Iterable[str, Operator]]): a target gateset for the device
```
Comment (albi3ro) [inline pennylane/devices/null_qubit.py:142]:
```suggestion
        operations (Optional[Iterable[str, type]]): a target gateset for the device
```

File: pennylane/devices/null_qubit.py [hunk 3]
```diff
@@ -137,52 +139,47 @@ class NullQubit(Device):
             (``['aux_wire', 'q1', 'q2']``). Default ``None`` if not specified.
         shots (int, Sequence[int], Sequence[Union[int, Sequence[int]]]): The default number of shots
             to use in executions involving this device.
+        operations (Optional[Iterable[str, type]]): a target gateset for the device.
+        target_device (Optional[pennylane.devices.Device]): a device to mimic the preprocessing of. Must obey the new device interface.
+        assume_no_broadcasting=False (bool): If ``True``, we always assume no parameter batching exists. Useful
+            for profiling and benchmarking.
```
Comment (obliviateandsurrender) [inline pennylane/devices/null_qubit.py:145]:
```suggestion
        assume_no_broadcasting (bool): If ``True``, we always assume no parameter batching exists. Useful
            for profiling and benchmarking. The default value is ``False`.
```

File: pennylane/devices/null_qubit.py [hunk 4]
```diff
@@ -137,52 +139,47 @@ class NullQubit(Device):
...
+            params = np.random.random(shape)
 
-        with qml.Tracker(dev) as tracker:
-            circuit(params)
+            with qml.Tracker(device) as tracker:
+                circuit(params)
+            return tracker
 
-    >>> tracker.history["resources"][0]
+    >>> track_entangling_layers(qml.device('null.qubit')).latest['resources']
```
Comment (obliviateandsurrender) [inline pennylane/devices/null_qubit.py:169]:
```suggestion
    >>> dev = qml.device('null.qubit')
    >>> track_entangling_layers(dev).latest['resources']
```
Looks more readable and less crowded this way.

File: pennylane/devices/null_qubit.py [hunk 5]
```diff
@@ -137,52 +139,47 @@ class NullQubit(Device):
...
+    shots: Shots(total=None)
+    gate_types:
+    {'StronglyEntanglingLayers': 1}
+    gate_sizes:
+    {100: 1}
+
+    Null qubit can also mimic the preprocessing of an existing device:
+
+    >>> target_dev = qml.device('default.qubit')
+    >>> track_entangling_layers(qml.device('null.qubit', target_device=target_dev)).latest['resources']
```
Comment (obliviateandsurrender) [inline pennylane/devices/null_qubit.py:182]:
```suggestion
    >>> dev = qml.device('null.qubit', target_device=target_dev)
    >>> track_entangling_layers(dev).latest['resources']
```

File: pennylane/devices/null_qubit.py [hunk 6]
```diff
@@ -192,6 +189,18 @@ def circuit(params):
     gate_sizes:
     {1: 5000, 2: 5000}
 
+    And decompose to a target gateset, specified by names and/or types:
+
+    >>> ops = {'RX', 'RY', 'RZ', qml.CNOT}
+    >>> track_entangling_layers(qml.device('null.qubit', operations=ops)).latest['resources']
```
Comment (obliviateandsurrender) [inline pennylane/devices/null_qubit.py:195]:
```suggestion
    >>> dev = qml.device('null.qubit', operations=ops)
    >>> track_entangling_layers(dev).latest['resources']
```

File: pennylane/devices/null_qubit.py [hunk 7]
```diff
@@ -221,15 +230,38 @@ def name(self):
...
+    def _get_batch_size(self, circuit):
+        return None if self._assume_no_broadcasting else circuit.batch_size
 
     def _simulate(self, circuit, interface):
         shots = circuit.shots
         obj_with_wires = self if self.wires else circuit
+
         results = tuple(
-            zero_measurement(mp, obj_with_wires, shots, circuit.batch_size, interface)
+            zero_measurement(mp, obj_with_wires, shots, self._get_batch_size(circuit), interface)
```
Comment (obliviateandsurrender) [inline pennylane/devices/null_qubit.py:264]: I remember you mentioning computing `batch_size` taking time? Should we then compute it just once and use it here for `results` and below for `derivatives`?

File: pennylane/devices/null_qubit.py [hunk 8]
```diff
@@ -280,7 +308,21 @@ def _max_workers(self):
     def preprocess(
         self, execution_config=DefaultExecutionConfig
     ) -> Tuple[TransformProgram, ExecutionConfig]:
-        program, _ = DefaultQubit.preprocess(self, execution_config)
+        if self._target_device:
+            return self._target_device.preprocess(execution_config)
+
+        program = TransformProgram()
```
Comment (obliviateandsurrender) [inline pennylane/devices/null_qubit.py:314]: Nice!

File: tests/devices/test_null_qubit.py [hunk 9]
```diff
@@ -110,6 +110,77 @@ def test_tracking():
...
+    assert len(program) == 1
+    assert program[0].transform == qml.defer_measurements.transform
+
+
+def test_target_device_and_operations():
+    """Test that a target device can be used together with specifying operations."""
+
+    # pylint: disable=unused-argument, too-few-public-methods
+    class DummyDev(qml.devices.Device):
+        """a device that just defers measurements."""
```
Comment (obliviateandsurrender) [inline tests/devices/test_null_qubit.py:166]:
```suggestion
        """A device that just defers measurements."""
```

File: tests/devices/test_null_qubit.py [hunk 10]
```diff
@@ -110,6 +110,77 @@ def test_tracking():
...
+    )
+    assert qml.equal(batch[0], expected)
+
+
+def test_target_device():
+    """Test that a target device can be provided to mimic."""
+
+    # pylint: disable=unused-argument
+    class DummyDev(qml.devices.Device):
+        """a device that just defers measurements."""
```
Comment (obliviateandsurrender) [inline tests/devices/test_null_qubit.py:144]:
```suggestion
        """A device that just defers measurements."""
```

File: tests/devices/test_null_qubit.py [hunk 11]
```diff
@@ -110,6 +110,77 @@ def test_tracking():
...
 
 
+def test_assume_no_broadcasting():
+    """Test that if no batch sizes are added if ``assume_no_broadcasting`` is True."""
+
+    dev = qml.device("null.qubit", assume_no_broadcasting=True)
+
+    tape = qml.tape.QuantumScript([qml.RX((1.2, 2.3), 0)], [qml.expval(qml.Z(0))])
+    results = dev.execute(tape)
+    assert qml.math.allclose(results, np.array(0.0))
```
Comment (obliviateandsurrender) [inline tests/devices/test_null_qubit.py:120]: How does this test no broadcasting? 🤔

File: tests/devices/test_null_qubit.py [hunk 12]
```diff
@@ -110,6 +110,77 @@ def test_tracking():
...
+    """Test that a target device can be used together with specifying operations."""
+
+    # pylint: disable=unused-argument, too-few-public-methods
+    class DummyDev(qml.devices.Device):
+        """a device that just defers measurements."""
+
+        def execute(self, circuits, execution_config=qml.devices.DefaultExecutionConfig):
+            return 0
+
+    with pytest.raises(NotImplementedError):
```
Comment (obliviateandsurrender) [inline tests/devices/test_null_qubit.py:171]: It might be nice to test the message as well.

**Review by obliviateandsurrender — COMMENTED** [review]
Thanks for the quick PR on this @albi3ro! Looking great, just leaving some comments and suggestions.

File: tests/devices/test_null_qubit.py (hunk 11 above)
Comment (albi3ro) [reply tests/devices/test_null_qubit.py:120]: the tape has a broadcast dimension.  So if it actually checked, the output would have an extra dimension.

**Review by trbromley — COMMENTED** [review]
Thanks @albi3ro! Are we still able to support the [usecase reported in `qml.resource`](https://xanaduai-pennylane--5527.com.readthedocs.build/en/5527/code/qml_resource.html#tracking-resources-for-custom-operations)?

It might be an idea to add that as a test if it's not there already.

The more I think about it, I think we need something like: `target_device` defaults to `default.qubit` _and_ we automatically support ops of type `ResourcesOperation`.

**Comment by trbromley** [comment]
@albi3ro shall we close this for now as it's replaced by https://github.com/PennyLaneAI/pennylane/pull/5582?

**Comment by albi3ro** [comment]
> @albi3ro shall we close this for now as it's replaced by #5582?

I'd still like to add the capability to specify a target device to `null.qubit`, but we can just create a new PR at a later point in time.

**Pull request closed without merging**